# Trip Management System — ERD Redesign, Implementation & Verification

**Exercise 2** — Data Engineering Practices

## Abstract

This notebook documents the redesign, physical implementation, and verification of the relational schema for **The Victory Leonard**, a provincial bus company's trip management system. An initial entity-relationship diagram (`ERD_LT1.png`) was presented and critiqued for three cardinality errors that, left uncorrected, would have caused lost maintenance history, an unrepresentable depot state, and an unrealistic staffing model. The corrected ERD (`ERD.md`) was implemented as a PostgreSQL schema on Amazon RDS, seeded with 377 synthetic rows across 8 tables (200 of them ticket transactions), and exercised end-to-end with CRUD operations. Each of the three flagged issues is shown, with live query evidence, to be resolved in the final schema.

## Motivation

Getting cardinality right in an ERD is not a cosmetic exercise — it determines what the database can and cannot represent once real data starts flowing through it. During the presentation of this system's initial ERD, three design choices were flagged as likely to cause real operational problems for a bus company:

- A **one-to-one bus↔maintenance** relation would silently overwrite service history every time a bus went in for maintenance again, destroying the audit trail that fleet managers and regulators rely on.
- A **many-to-zero depot↔bus** notation had the cardinality backwards: a newly opened depot legitimately starts out with zero buses, but the diagram implied a depot could only exist once it already owned buses.
- A **one-to-one employee↔schedule** relation would prevent any driver or conductor from ever working more than one trip, which contradicts how bus crews are actually staffed.

Left uncorrected, any one of these would either lose data silently or block the system from modeling real fleet operations. This notebook works through the fix for each issue, then proves the fix with live data rather than just the diagram.

## Introduction

**The Victory Leonard** trip management system tracks the operational backbone of a provincial bus company: which buses are housed at which depots, when each bus was last serviced, which routes and stops exist, which employees are staffed as drivers or conductors, which dated trips (`schedule`) are running, and which tickets were sold against them. The design work below starts from an initial ERD that was presented and critiqued, then follows that feedback through to a corrected, implemented, and verified schema.

## Definition of Terms

**Modeling terms**

- **ERD (Entity-Relationship Diagram):** a diagram describing the tables (entities), their attributes, and how they relate to one another.
- **Primary key (PK):** the column that uniquely identifies a row in a table.
- **Foreign key (FK):** a column that references a primary key in another table, enforcing that a relationship must point to a row that actually exists.
- **Cardinality:** how many rows on one side of a relationship may correspond to rows on the other side — e.g. *one-to-one*, *one-to-many*, *zero-or-many*. Written here in Mermaid/crow's-foot notation as `||` (exactly one) and `o{` (zero or many).
- **Normalization:** structuring tables so each fact is stored once, avoiding the kind of overwrite/anomaly problems described under Motivation.
- **Idempotent (script):** safe to re-run — e.g. the schema rebuild below `DROP`s tables before `CREATE`-ing them, so running it twice produces the same end state.
- **CRUD:** the four basic data operations — **C**reate, **R**ead, **U**pdate, **D**elete — used here to exercise the corrected schema end-to-end.

**Domain terms**

- **Depot:** a physical facility that houses buses when they are not in service.
- **Route:** a defined path between a start and end point, made up of an ordered sequence of stops.
- **Stop:** a named point along a route where passengers may board or alight.
- **Schedule:** one dated, timed trip — a specific bus, driver, and conductor assigned to run a route between a departure and arrival time.
- **Ticket:** a single passenger's paid trip transaction, tied to one schedule and a source/destination stop pair.
- **Manifest:** the set of tickets sold against a single schedule, i.e. who is on that particular trip.
- **Maintenance record:** one dated service event for a bus (`completed`, `ongoing`, or `scheduled`).

## Methodology

With the issues above scoped out, the redesign was carried out in the following steps: (1) provision and connect to a managed Postgres instance on Amazon RDS, (2) review the initial ERD and the feedback it received, (3) redesign the schema to resolve the flagged cardinality issues, (4) implement the corrected schema as DDL, (5) seed it with synthetic data, (6) verify the seeded data through row counts, column schemas, and sample rows, and (7) exercise the schema with CRUD operations that directly demonstrate each fix working as intended.

### RDS Provisioning & Connection

The database runs on a managed PostgreSQL instance on Amazon RDS, configured as shown below. The notebook connects to it via `jupysql`, with credentials loaded from a local, git-ignored `pw.txt` (see [`setup_admin.sql`](setup_admin.sql) for how the application role and database were bootstrapped).

![RDS_Ex2.png](RDS_Ex2.png)

In [1]:
%load_ext sql

In [2]:
import os

with open('pw.txt') as f:
    conn_string = f.read()

os.environ['DATABASE_URL'] = conn_string

In [3]:
%sql \l

Name,Owner,Encoding,Collate,Ctype,Access privileges
postgres,postgres,UTF8,en_US.UTF-8,en_US.UTF-8,None
rdsadmin,rdsadmin,UTF8,en_US.UTF-8,en_US.UTF-8,rdsadmin=CTc/rdsadmin
template0,rdsadmin,UTF8,en_US.UTF-8,en_US.UTF-8,=c/rdsadminrdsadmin=CTc/rdsadmin
template1,postgres,UTF8,en_US.UTF-8,en_US.UTF-8,=c/postgrespostgres=CTc/postgres


### Reviewing the Initial ERD

The ERD below (`ERD_LT1.png`) is the design as originally presented for The Victory Leonard's trip management system.

![ERD_LT1.png](ERD_LT1.png)

#### Issues Identified

- **Bus ↔ Maintenance** is drawn one-to-one, which overwrites past maintenance history on every new service record.
- **Depot ↔ Bus** is tagged many-to-zero — a notation error; it should read zero-to-many, since a newly built depot can go without buses for a while.
- **Employee ↔ Schedule** reads one-to-one, which is unrealistic: bus companies staff employees across flexible, recurring schedules, so the relation should be one-to-many.

### Schema Redesign

Resolving each issue identified above in the physical schema:

1. **Maintenance history preserved.** `maintenance` is now one-to-many with `bus`: every service event is its own row (`maintenance_id` PK, `maintenance_date`, `next_maintenance_due`) instead of a single per-bus record that got overwritten. The misleading `insurance_id` PK is renamed to `maintenance_id`.
2. **Depot–Bus cardinality corrected.** `bus.depot_id` is a nullable-side FK from the *bus* to its depot, so one depot houses **zero or many** buses (a newly opened depot can have none yet).
3. **Employee–Schedule made one-to-many.** `schedule` now stores an **actual dated trip** (`departure_time` / `arrival_time` timestamps, not a recurring template), with two role FKs — `driver_id` and `conductor_id` — so one employee can work many schedules over time.

**Additional design decision:** `ticket.schedule_id` was added. In the original ERD a ticket referenced only its source/destination stops, so it wasn't tied to any particular trip. Linking each ticket to a dated `schedule` makes it a true *trip transaction* and enables per-trip manifests and revenue reporting.

### Implementing the Corrected Schema (DDL)

The schema is rebuilt idempotently — children are dropped before parents — then recreated with the fixes above applied as constraints (nullable vs. `NOT NULL` foreign keys, `CHECK` constraints, and the new `maintenance` / `schedule` / `ticket` shapes).

In [4]:
%%sql
-- idempotent rebuild: drop children before parents
DROP TABLE IF EXISTS ticket, schedule, maintenance, stops, route, bus,
    employee, depot CASCADE;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

++
||
++
++

In [5]:
%%sql
CREATE TABLE depot (
    depot_id   SERIAL PRIMARY KEY,
    depot_name VARCHAR(100) NOT NULL
);

CREATE TABLE bus (
    bus_id    SERIAL PRIMARY KEY,
    bus_name  VARCHAR(100) NOT NULL,
    model     VARCHAR(100),
    capacity  INT NOT NULL CHECK (capacity > 0),
    ownership VARCHAR(50),
    -- fix #2: a depot houses zero-or-many buses
    depot_id  INT REFERENCES depot (depot_id)
);

-- fix #1: one bus -> many maintenance records (history preserved)
CREATE TABLE maintenance (
    maintenance_id       SERIAL PRIMARY KEY,
    bus_id               INT NOT NULL REFERENCES bus (bus_id),
    status               VARCHAR(30) NOT NULL
        CHECK (status IN ('completed', 'ongoing', 'scheduled')),
    maintenance_date     DATE NOT NULL,
    next_maintenance_due DATE
);

CREATE TABLE route (
    route_id    SERIAL PRIMARY KEY,
    start_point VARCHAR(100) NOT NULL,
    stop_point  VARCHAR(100) NOT NULL,
    distance    NUMERIC(7, 1) NOT NULL,  -- km
    no_of_stops INT NOT NULL
);

CREATE TABLE stops (
    stop_id   SERIAL PRIMARY KEY,
    stop_name VARCHAR(100) NOT NULL,
    route_id  INT NOT NULL REFERENCES route (route_id)
);

CREATE TABLE employee (
    emp_id     SERIAL PRIMARY KEY,
    first_name VARCHAR(50) NOT NULL,
    last_name  VARCHAR(50) NOT NULL,
    type       VARCHAR(20) NOT NULL CHECK (type IN ('driver', 'conductor')),
    salary     NUMERIC(10, 2) NOT NULL
);

-- fix #3: a schedule is an actual dated trip, so one employee -> many
-- schedules
CREATE TABLE schedule (
    schedule_id    SERIAL PRIMARY KEY,
    route_id       INT NOT NULL REFERENCES route (route_id),
    bus_id         INT NOT NULL REFERENCES bus (bus_id),
    driver_id      INT NOT NULL REFERENCES employee (emp_id),
    conductor_id   INT NOT NULL REFERENCES employee (emp_id),
    departure_time TIMESTAMP NOT NULL,
    arrival_time   TIMESTAMP NOT NULL,
    CHECK (arrival_time > departure_time)
);

-- ticket now references its dated trip (schedule_id): a true trip transaction
CREATE TABLE ticket (
    ticket_id      SERIAL PRIMARY KEY,
    schedule_id    INT NOT NULL REFERENCES schedule (schedule_id),
    source_id      INT NOT NULL REFERENCES stops (stop_id),
    destination_id INT NOT NULL REFERENCES stops (stop_id),
    fare           NUMERIC(8, 2) NOT NULL CHECK (fare >= 0),
    CHECK (source_id <> destination_id)
);

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

++
||
++
++

### Updated ERD

Corrected ERD (also saved as [`ERD.md`](ERD.md)):

```mermaid
erDiagram
    DEPOT {
        int depot_id PK
        varchar depot_name
    }

    BUS {
        int bus_id PK
        varchar bus_name
        varchar model
        int capacity
        varchar ownership
        int depot_id FK
    }

    MAINTENANCE {
        int maintenance_id PK
        int bus_id FK
        varchar status
        date maintenance_date
        date next_maintenance_due
    }

    ROUTE {
        int route_id PK
        varchar start_point
        varchar stop_point
        numeric distance
        int no_of_stops
    }

    STOPS {
        int stop_id PK
        varchar stop_name
        int route_id FK
    }

    EMPLOYEE {
        int emp_id PK
        varchar first_name
        varchar last_name
        varchar type
        numeric salary
    }

    SCHEDULE {
        int schedule_id PK
        int route_id FK
        int bus_id FK
        int driver_id FK
        int conductor_id FK
        timestamp departure_time
        timestamp arrival_time
    }

    TICKET {
        int ticket_id PK
        int schedule_id FK
        int source_id FK
        int destination_id FK
        numeric fare
    }

    DEPOT ||--o{ BUS : "houses"
    BUS ||--o{ MAINTENANCE : "has history of"
    ROUTE ||--o{ STOPS : "consists of"
    ROUTE ||--o{ SCHEDULE : "is served by"
    BUS ||--o{ SCHEDULE : "is assigned to"
    EMPLOYEE ||--o{ SCHEDULE : "drives (driver_id)"
    EMPLOYEE ||--o{ SCHEDULE : "conducts (conductor_id)"
    SCHEDULE ||--o{ TICKET : "sells"
    STOPS ||--o{ TICKET : "is source of"
    STOPS ||--o{ TICKET : "is destination of"
```

### Data Seeding

Synthetic data is loaded from [`seed_data.sql`](seed_data.sql) — a Philippine provincial-bus-themed dataset (depots, buses, routes/stops, employees, dated schedules, and ticket transactions) sized to plausibly exercise every relationship in the schema, including the zero-bus depot and multi-trip drivers called out above.

In [7]:
import psycopg2

with open('seed_data.sql') as f:
    seed_sql = f.read()

conn = psycopg2.connect(os.environ['DATABASE_URL'])
with conn, conn.cursor() as cur:
    cur.execute(seed_sql)
conn.close()
print('seed_data.sql applied')

seed_data.sql applied


### Verification

Before exercising CRUD operations, the seeded data is verified through row counts, the table/column inventory, and a sample of rows from each table.

#### Sanity Checks

In [8]:
%%sql
-- sanity check: row counts per table (ticket must be 200)
SELECT 'depot' AS table_name, COUNT(*) AS n_rows FROM depot
UNION ALL SELECT 'bus',         COUNT(*) FROM bus
UNION ALL SELECT 'maintenance', COUNT(*) FROM maintenance
UNION ALL SELECT 'route',       COUNT(*) FROM route
UNION ALL SELECT 'stops',       COUNT(*) FROM stops
UNION ALL SELECT 'employee',    COUNT(*) FROM employee
UNION ALL SELECT 'schedule',    COUNT(*) FROM schedule
UNION ALL SELECT 'ticket',      COUNT(*) FROM ticket
ORDER BY table_name;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

8 rows affected.

table_name,n_rows
bus,14
depot,5
employee,24
maintenance,42
route,10
schedule,40
stops,42
ticket,200


#### Table Inventory & Column Schema

Listing the tables, then the per-table column schema (row counts already shown above).

In [9]:
%sql \dt

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

Schema,Name,Type,Owner
public,bus,table,postgres
public,depot,table,postgres
public,employee,table,postgres
public,maintenance,table,postgres
public,route,table,postgres
public,schedule,table,postgres
public,stops,table,postgres
public,ticket,table,postgres


*(`\d table_name` hits a known jupysql/pgspecial incompatibility — `\d` returns multiple result sets as a generator that jupysql's `ResultSet` can't slice — so `%sqlcmd columns`, the jupysql-native equivalent, is used instead.)*

In [10]:
%sqlcmd columns -t depot

name,type,nullable,default,autoincrement,comment
depot_id,INTEGER,False,nextval('depot_depot_id_seq'::regclass),True,None
depot_name,VARCHAR(100),False,None,False,None


In [11]:
%sqlcmd columns -t bus

name,type,nullable,default,autoincrement,comment
bus_id,INTEGER,False,nextval('bus_bus_id_seq'::regclass),True,None
bus_name,VARCHAR(100),False,None,False,None
model,VARCHAR(100),True,None,False,None
capacity,INTEGER,False,None,False,None
ownership,VARCHAR(50),True,None,False,None
depot_id,INTEGER,True,None,False,None


In [12]:
%sqlcmd columns -t maintenance

name,type,nullable,default,autoincrement,comment
maintenance_id,INTEGER,False,nextval('maintenance_maintenance_id_seq'::regclass),True,None
bus_id,INTEGER,False,None,False,None
status,VARCHAR(30),False,None,False,None
maintenance_date,DATE,False,None,False,None
next_maintenance_due,DATE,True,None,False,None


In [13]:
%sqlcmd columns -t route

name,type,nullable,default,autoincrement,comment
route_id,INTEGER,False,nextval('route_route_id_seq'::regclass),True,None
start_point,VARCHAR(100),False,None,False,None
stop_point,VARCHAR(100),False,None,False,None
distance,"NUMERIC(7, 1)",False,None,False,None
no_of_stops,INTEGER,False,None,False,None


In [14]:
%sqlcmd columns -t stops

name,type,nullable,default,autoincrement,comment
stop_id,INTEGER,False,nextval('stops_stop_id_seq'::regclass),True,None
stop_name,VARCHAR(100),False,None,False,None
route_id,INTEGER,False,None,False,None


In [15]:
%sqlcmd columns -t employee

name,type,nullable,default,autoincrement,comment
emp_id,INTEGER,False,nextval('employee_emp_id_seq'::regclass),True,None
first_name,VARCHAR(50),False,None,False,None
last_name,VARCHAR(50),False,None,False,None
type,VARCHAR(20),False,None,False,None
salary,"NUMERIC(10, 2)",False,None,False,None


In [16]:
%sqlcmd columns -t schedule

name,type,nullable,default,autoincrement,comment
schedule_id,INTEGER,False,nextval('schedule_schedule_id_seq'::regclass),True,None
route_id,INTEGER,False,None,False,None
bus_id,INTEGER,False,None,False,None
driver_id,INTEGER,False,None,False,None
conductor_id,INTEGER,False,None,False,None
departure_time,TIMESTAMP,False,None,False,None
arrival_time,TIMESTAMP,False,None,False,None


In [17]:
%sqlcmd columns -t ticket

name,type,nullable,default,autoincrement,comment
ticket_id,INTEGER,False,nextval('ticket_ticket_id_seq'::regclass),True,None
schedule_id,INTEGER,False,None,False,None
source_id,INTEGER,False,None,False,None
destination_id,INTEGER,False,None,False,None
fare,"NUMERIC(8, 2)",False,None,False,None


#### Sample Data (First 100 Rows per Table)

All 8 tables have fewer than 100 rows except `ticket` (200 rows), so every query below returns its table in full except the ticket query, which is capped at the first 100 by `ticket_id`.

In [18]:
%%sql
-- first 100 rows: depot
SELECT * FROM depot ORDER BY depot_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

5 rows affected.

depot_id,depot_name
1,Cubao Depot
2,PITX Depot
3,Buendia Depot
4,Sampaloc Depot
5,Sta. Rosa Depot (new)


In [19]:
%%sql
-- first 100 rows: bus
SELECT * FROM bus ORDER BY bus_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

14 rows affected.

bus_id,bus_name,model,capacity,ownership,depot_id
1,VL-101,Volvo 9700,45,company-owned,3
2,VL-102,King Long XMQ6117Y,49,company-owned,1
3,VL-103,Volvo 9700,45,leased,4
4,VL-104,Yutong ZK6122H9,45,company-owned,2
5,VL-105,King Long XMQ6117Y,45,leased,2
6,VL-106,Volvo 9700,60,company-owned,4
7,VL-107,Golden Dragon XML6127,55,company-owned,2
8,VL-108,Volvo 9700,60,company-owned,3
9,VL-109,King Long XMQ6117Y,49,company-owned,1
10,VL-110,Yutong ZK6122H9,60,company-owned,3


In [20]:
%%sql
-- first 100 rows: maintenance
SELECT * FROM maintenance ORDER BY maintenance_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

42 rows affected.

maintenance_id,bus_id,status,maintenance_date,next_maintenance_due
1,1,completed,2025-09-08,2025-12-07
2,1,completed,2025-11-22,2026-02-20
3,1,completed,2026-03-08,2026-06-06
4,1,ongoing,2026-06-10,2026-09-08
5,2,completed,2025-10-31,2026-01-29
6,2,completed,2026-02-08,2026-05-09
7,2,scheduled,2026-06-07,2026-09-05
8,3,completed,2026-03-06,2026-06-04
9,3,completed,2026-05-25,2026-08-23
10,4,completed,2026-03-19,2026-06-17


In [21]:
%%sql
-- first 100 rows: route
SELECT * FROM route ORDER BY route_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

10 rows affected.

route_id,start_point,stop_point,distance,no_of_stops
1,Manila,Baguio,250.0,4
2,Cubao,Baguio,246.0,4
3,Cubao,Boracay (Caticlan),690.0,5
4,Buendia,Davao,1450.0,5
5,PITX,Lucena,130.0,3
6,PITX,Batangas,110.0,3
7,Manila,Vigan,400.0,4
8,Cubao,Tuguegarao,480.0,5
9,Buendia,Naga,380.0,4
10,PITX,Iloilo,700.0,5


In [22]:
%%sql
-- first 100 rows: stops
SELECT * FROM stops ORDER BY stop_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

42 rows affected.

stop_id,stop_name,route_id
1,Manila,1
2,Tarlac City,1
3,Rosario,1
4,Baguio,1
5,Cubao,2
6,San Fernando (Pampanga),2
7,Urdaneta,2
8,Baguio,2
9,Cubao,3
10,Batangas Port,3


In [23]:
%%sql
-- first 100 rows: employee
SELECT * FROM employee ORDER BY emp_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

24 rows affected.

emp_id,first_name,last_name,type,salary
1,Jose,Santos,driver,28375.00
2,Maria,Reyes,driver,33417.00
3,Juan,Cruz,driver,29866.00
4,Ana,Bautista,driver,34332.00
5,Pedro,Garcia,driver,30370.00
6,Liza,Mendoza,driver,28653.00
7,Ramon,Torres,driver,29907.00
8,Grace,Flores,driver,28827.00
9,Carlo,Ramos,driver,31113.00
10,Nena,Gonzales,driver,30277.00


In [24]:
%%sql
-- first 100 rows: schedule
SELECT * FROM schedule ORDER BY schedule_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

40 rows affected.

schedule_id,route_id,bus_id,driver_id,conductor_id,departure_time,arrival_time
1,6,3,8,13,2026-06-08 14:45:00,2026-06-08 16:45:00
2,5,14,11,17,2026-07-03 10:00:00,2026-07-03 12:21:49.090909
3,9,13,3,21,2026-06-13 09:30:00,2026-06-13 16:24:32.727273
4,9,1,2,18,2026-06-01 15:45:00,2026-06-01 22:39:32.727273
5,5,10,2,14,2026-06-16 06:15:00,2026-06-16 08:36:49.090909
6,8,3,11,20,2026-06-05 22:15:00,2026-06-06 06:58:38.181818
7,9,4,9,24,2026-06-11 13:45:00,2026-06-11 20:39:32.727273
8,4,8,9,20,2026-06-20 17:30:00,2026-06-21 19:51:49.090909
9,2,6,1,22,2026-06-16 12:00:00,2026-06-16 16:28:21.818182
10,9,2,12,23,2026-06-15 12:00:00,2026-06-15 18:54:32.727273


In [25]:
%%sql
-- first 100 rows: ticket (200 total, showing first 100 by ticket_id)
SELECT * FROM ticket ORDER BY ticket_id LIMIT 100;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

100 rows affected.

ticket_id,schedule_id,source_id,destination_id,fare
1,27,14,15,847.50
2,22,3,4,233.33
3,16,15,16,847.50
4,7,34,37,886.00
5,31,34,35,328.67
6,30,23,24,171.00
7,15,1,2,233.33
8,13,40,41,435.00
9,18,1,2,233.33
10,23,32,33,314.00


### CRUD Verification

Exercising the corrected schema with the four CRUD verbs, plus one query per schema fix proving the redesign behaves as intended.

In [26]:
%%sql
-- READ: per-trip manifest — route, bus, crew, tickets sold, revenue
-- (top 10 by revenue)
SELECT s.schedule_id,
       r.start_point || ' → ' || r.stop_point AS route,
       b.bus_name,
       d.first_name || ' ' || d.last_name     AS driver,
       s.departure_time,
       COUNT(t.ticket_id)                     AS tickets_sold,
       COALESCE(SUM(t.fare), 0)               AS revenue_php
FROM schedule s
JOIN route    r ON r.route_id = s.route_id
JOIN bus      b ON b.bus_id   = s.bus_id
JOIN employee d ON d.emp_id   = s.driver_id
LEFT JOIN ticket t ON t.schedule_id = s.schedule_id
GROUP BY s.schedule_id, route, b.bus_name, driver, s.departure_time
ORDER BY revenue_php DESC
LIMIT 10;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

10 rows affected.

schedule_id,route,bus_name,driver,departure_time,tickets_sold,revenue_php
16,Buendia → Davao,VL-103,Ramon Torres,2026-06-13 11:45:00,9,11615.00
27,Buendia → Davao,VL-111,Efren Villanueva,2026-06-17 17:15:00,8,11565.00
25,Buendia → Davao,VL-111,Nena Gonzales,2026-06-03 07:45:00,8,10767.50
8,Buendia → Davao,VL-108,Carlo Ramos,2026-06-20 17:30:00,4,7377.50
33,Cubao → Boracay (Caticlan),VL-112,Liza Mendoza,2026-06-18 14:15:00,6,5233.50
31,Buendia → Naga,VL-111,Carlo Ramos,2026-06-20 21:00:00,9,4908.66
21,Manila → Vigan,VL-104,Pedro Garcia,2026-07-06 20:15:00,9,4849.99
38,Buendia → Naga,VL-101,Pedro Garcia,2026-06-10 18:15:00,7,4251.33
13,PITX → Iloilo,VL-113,Ramon Torres,2026-07-01 12:45:00,7,4200.00
7,Buendia → Naga,VL-104,Carlo Ramos,2026-06-11 13:45:00,5,3594.00


In [27]:
%%sql
-- fix #1 proof: full maintenance history retained per bus (was
-- overwritten under 1-to-1)
SELECT m.maintenance_id, b.bus_name, m.status, m.maintenance_date,
       m.next_maintenance_due
FROM maintenance m
JOIN bus b USING (bus_id)
WHERE b.bus_id = 1
ORDER BY m.maintenance_date;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

4 rows affected.

maintenance_id,bus_name,status,maintenance_date,next_maintenance_due
1,VL-101,completed,2025-09-08,2025-12-07
2,VL-101,completed,2025-11-22,2026-02-20
3,VL-101,completed,2026-03-08,2026-06-06
4,VL-101,ongoing,2026-06-10,2026-09-08


In [28]:
%%sql
-- fix #2 proof: a depot may hold zero buses (Sta. Rosa is newly opened)
SELECT d.depot_name, COUNT(b.bus_id) AS buses_housed
FROM depot d
LEFT JOIN bus b USING (depot_id)
GROUP BY d.depot_name
ORDER BY buses_housed DESC;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

5 rows affected.

depot_name,buses_housed
Cubao Depot,4
Buendia Depot,4
Sampaloc Depot,3
PITX Depot,3
Sta. Rosa Depot (new),0


In [29]:
%%sql
-- fix #3 proof: one employee works many dated schedules (was 1-to-1)
SELECT e.first_name || ' ' || e.last_name AS driver, COUNT(*) AS trips_assigned
FROM schedule s
JOIN employee e ON e.emp_id = s.driver_id
GROUP BY e.emp_id, driver
ORDER BY trips_assigned DESC
LIMIT 5;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

5 rows affected.

driver,trips_assigned
Carlo Ramos,6
Liza Mendoza,4
Juan Cruz,4
Jose Santos,4
Maria Reyes,4


In [30]:
%%sql
-- CREATE: a walk-in passenger buys an end-to-end ticket on schedule 1
INSERT INTO ticket (schedule_id, source_id, destination_id, fare)
SELECT s.schedule_id, MIN(st.stop_id), MAX(st.stop_id), 50 + r.distance * 2.2
FROM schedule s
JOIN route r  ON r.route_id = s.route_id
JOIN stops st ON st.route_id = s.route_id
WHERE s.schedule_id = 1
GROUP BY s.schedule_id, r.distance
RETURNING ticket_id, schedule_id, source_id, destination_id, fare;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

1 rows affected.

ticket_id,schedule_id,source_id,destination_id,fare
201,1,22,24,292.00


In [31]:
%%sql
-- UPDATE: ongoing maintenance jobs wrap up -> completed (the history
-- rows remain)
UPDATE maintenance
SET status = 'completed'
WHERE status = 'ongoing'
RETURNING maintenance_id, bus_id, status, maintenance_date;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

4 rows affected.

maintenance_id,bus_id,status,maintenance_date
4,1,completed,2026-06-10
11,4,completed,2026-05-20
28,9,completed,2026-06-04
42,14,completed,2026-05-23


In [32]:
%%sql
-- DELETE: the walk-in passenger asks for a refund; remove the ticket
-- created above
DELETE FROM ticket
WHERE ticket_id = (SELECT MAX(ticket_id) FROM ticket)
RETURNING ticket_id, schedule_id, fare AS refunded_fare;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

1 rows affected.

ticket_id,schedule_id,refunded_fare
201,1,292.00


In [33]:
%%sql
-- final state: the 200 synthesized trip transactions are intact
SELECT COUNT(*) AS total_trip_transactions FROM ticket;

Running query in 'postgresql://postgres:***@sampledb0.c6rcg6kqa73f.us-east-1.rds.amazonaws.com/postgres'

1 rows affected.

total_trip_transactions
200


## Finding

**Seed data integrity.** The sanity-check row counts (`depot`=5, `bus`=14, `maintenance`=42, `route`=10, `stops`=42, `employee`=24, `schedule`=40, `ticket`=200 — 377 rows total) match the seed script exactly, confirming the schema accepted every relationship the synthetic dataset exercises without constraint violations.

**Trip manifests are queryable end-to-end.** Joining `schedule` → `route`/`bus`/`employee` → `ticket` produces a per-trip manifest with ticket counts and revenue; the Buendia → Davao route dominates the top 10 by revenue (up to ₱11,615 on a single trip), which is expected given it is the longest route (1,450 km) in the dataset.

**Fix #1 confirmed — maintenance history is preserved.** Bus VL-101 now carries 4 distinct maintenance rows spanning 2025-09-08 through 2026-06-10, instead of a single row that would have been overwritten under the original one-to-one design.

**Fix #2 confirmed — a depot can hold zero buses.** Of the 5 depots, Sta. Rosa Depot (newly opened) correctly shows 0 buses housed, which the original many-to-zero notation could not represent.

**Fix #3 confirmed — employees work many schedules.** Driver Carlo Ramos is assigned to 6 separate dated trips, and four other drivers are assigned to 4 each — impossible under the original one-to-one employee↔schedule design.

**CRUD operations behave correctly.** A walk-in ticket was created end-to-end on schedule 1 (`ticket_id` 201), four `ongoing` maintenance jobs were transitioned to `completed` without losing their history rows, and the walk-in ticket was deleted on refund — leaving the `ticket` table back at its original 200 rows.

**Conclusion.** All three cardinality issues raised during the initial ERD's presentation were corrected in the physical schema and independently verified with live queries against seeded data, not just asserted in the diagram.